### PR TITLE
KEYCLOAK-13950 SAML2 Identity Provider - Send Subject in SAML requests

### PR DESCRIPTION
**Pull request by looorent**
## Motivation

When working with SAML IDPs, it is currently not possible to send a Subject. This subject can prefill the username input field during the login process.

As mentionned by @hmlnarik in [KEYCLOAK-13858](https://issues.redhat.com/browse/KEYCLOAK-13858), the Subject is an optional standard element of SAML requests: See https://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf, line 2017 and below.

Another motivation: this is impossible to achieve this using `SamlAuthenticationPreprocessor` because:
* There is nothing from the request url available in the `clientSession`
* `SAMLRequestWriter` does not write the subject in the request

## Solution

This PR takes a query parameter named `login_hint` into account and add it to the SAML Request as a Subject element. When no login_hint is provided, no Subject is added to the request.

This does not require any change in the identity provider configuration.

## Limitations

Pay attention that several IDPs do not support the Subject element of SAML. For instance: ADFS simply ignores it: https://docs.microsoft.com/en-us/azure/active-directory/develop/single-sign-on-saml-protocol#subject

## `login_hint`

Why a `login_hint` and not a `subject` query parameter? I believe Keycloak exposes its Identity Providers (SAML or not) as an OpenID facade and then redirect the requests to the appropriate providers.
Considering OpenID already has a standard `login_hint` query parameter, this makes sense to support the same contract.

## Usage

By default, a social button pointing to a SAML Identity Provider redirects the user to a login URL `http[s]://{host:port}/auth/realms/{realm-name}/broker/{broker-alias}/login?<query-parameters>`
Adding a `login_hint` query parameter to this URL will add its value to SAML request as a Subject attribute.

### CODE REVIEW

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 1]
```diff
@@ -91,13 +93,15 @@ public Response performLogin(AuthenticationRequest request) {
                 protocolBinding = JBossSAMLURIConstants.SAML_HTTP_POST_BINDING.get();
             }
 
+            String loginHint = request.getHttpRequest().getUri().getQueryParameters().getFirst(LOGIN_HINT_QUERY_PARAMETER);
             SAML2AuthnRequestBuilder authnRequestBuilder = new SAML2AuthnRequestBuilder()
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:97]: Is there some input validation / escaping required here?

**Comment by thomasdarimont** [comment]
Great feature! had to build something like that as well, but for ADFS...

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java (hunk 1 above)
Comment (looorent) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:97]: This value is sanitized in the `subject` method of `SAML2AuthnRequestBuilder`. It is also escaped using Stax when it is written in the SAML request. Do you think something else is required?

**Comment by looorent** [comment]
> Great feature! had to build something like that as well, but for ADFS...

ADFS does not support it, indeed. However, I wrote a preprocessor that should inject `username` and `login_hint` in the destination URL. You can use it in your own keycloak install if you want. https://github.com/looorent/keycloak-saml-adfs-adapter This requires [KEYCLOAK-13962](https://issues.redhat.com/browse/KEYCLOAK-13962) to be accepted and merged first.

**Comment by sventorben** [comment]
What is current status on this one? Would love to see this available soon.

**Comment by hmlnarik** [comment]
Please check Travis logs. The failures are relevant.

**Comment by looorent** [comment]
Thanks @hmlnarik.
The tests are now ok.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 2]
```diff
@@ -91,13 +93,15 @@ public Response performLogin(AuthenticationRequest request) {
                 protocolBinding = JBossSAMLURIConstants.SAML_HTTP_POST_BINDING.get();
             }
 
+            String loginHint = request.getHttpRequest().getUri().getQueryParameters().getFirst(LOGIN_HINT_QUERY_PARAMETER);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:96]: This needs to be guarded by a configuration that mimics OIDC identity provider `loginHint` configuration which will be disabled by default:

1. Move `OAuth2IdentityProviderConfig`'s `isLoginHint` and `setLoginHint` to IdentityProviderModel
2. Add a `QueryParam(OIDCLoginProtocol.LOGIN_HINT_PARAM)` to authentication session in `IdentityBrokerService.performLogin` similarly to https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java#L445
3. In this place, only use `request.getAuthenticationSession().getClientNote(OIDCLoginProtocol.LOGIN_HINT_PARAM)`

Also update `realm-identity-provider-saml.html` to configure loginHint similarly to `realm-identity-provider-oidc.html`, and ensure that the label and label hint are properly adjusted to SAML terms.

The tests need to check that the settings of the loginHint configuration in SAML identity provider respect the settings.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 3]
```diff
@@ -59,6 +59,8 @@
  */
 public class SAMLIdentityProvider extends AbstractIdentityProvider<SAMLIdentityProviderConfig> {
     protected static final Logger logger = Logger.getLogger(SAMLIdentityProvider.class);
+    private static final String LOGIN_HINT_QUERY_PARAMETER = "login_hint";
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:62]: Please remove, use existing constant instead (see below)

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you for the contribution. The functionality can be misused, yet I believe this is a useful addition for certain cases. For that reason, the loginHint feature must be explicitly enabled for a particular identity provider before can be used.

Good news is that the OAuth2 identity provider already contains all the necessary bits, so the changes would be only matter of generalizing it and updating the SAML configuration page. Please see inline

**Comment by looorent** [comment]
Thanks @hmlnarik for your comments.
I have updated this PR to:
* pass `login_hint` through the session notes
* define a new option "loginHint" on SAML providers (called "Pass subject" in the admin web interface)
* write some tests assuming a boolean "loginHint" attribute on providers

The documentation has been updated accordingly (https://github.com/keycloak/keycloak-documentation/pull/923).

**Review by hmlnarik — APPROVED** [review]

**Pull request merged**